internal/gcs/container.go

```diff
@@ -19,15 +22,34 @@ const hrComputeSystemDoesNotExist = 0xc037010e
 // Container implements the cow.Container interface for containers
 // created via GuestConnection.
 type Container struct {
-	gc        *GuestConnection
-	id        string
-	notifyCh  chan struct{}
-	closeCh   chan struct{}
+	gc *GuestConnection
+	id string
+	// Blocks Wait() until waitBackground() returns or Close() is called.
+	// Does not represent the status of the container in the guest
+	waitBlock chan struct{}
+	// set only in closeOnce, read after <-waitBlock returns
+	waitError error
 	closeOnce sync.Once
+	// The channel notifications are received on from the guest connection
+	notifyCh chan notifications.Message
+	// Closed by GuestConnection to signal container in the guest existed
+	closeCh       chan struct{}
+	notifications *queue.MessageQueue
```

Comment:
This scares me a teeny bit but I know most of it was necessary for how you implemented the notifications. You ran all of the cri-containerd tests multiple times with this I take it? This is a pretty critical code path and even though what we had here might have had some dead code and oddities it hasn't caused any problems for quite awhile now. This is definitely for the better but I wonder how much (if any) of this we could move to its own PR.

Reply:
I put this comment on a weird bit of code, I'm mainly referring to the restructuring of most of the waiting and closing throughout this file